crypto/src/merkle/proofs.rs

```diff
@@ -255,6 +255,158 @@ impl<H: Hasher> BatchMerkleProof<H> {
+                            }
+                            buf[1] = self.leaves[index2];
+                        }
+                        None => return Err(MerkleTreeError::InvalidProof),
+                    }
+                    proof_pointers.push(1);
+                }
+            }
+
+            // hash sibling nodes into their parent and add it to partial_tree
+            let parent = H::merge(&buf);
+            partial_tree_map.insert(offset + index, buf[0]);
+            partial_tree_map.insert((offset + index) ^ 1, buf[1]);
+            let parent_index = (offset + index) >> 1;
+            v.insert(parent_index, parent);
+            next_indexes.push(parent_index);
+            partial_tree_map.insert(parent_index, parent);
+        }
+
+        // iteratively move up, until we get to the root
+        for _ in 1..self.depth {
+            let indexes = next_indexes.clone();
+            next_indexes.clear();
+
+            let mut i = 0;
+            while i < indexes.len() {
+                let node_index = indexes[i];
+                let sibling_index = node_index ^ 1;
+
+                // determine the sibling
+                let sibling = if i + 1 < indexes.len() && indexes[i + 1] == sibling_index {
+                    i += 1;
+                    match v.get(&sibling_index) {
+                        Some(sibling) => *sibling,
+                        None => return Err(MerkleTreeError::InvalidProof),
+                    }
+                } else {
+                    let pointer = proof_pointers[i];
+                    if self.nodes[i].len() <= pointer {
+                        return Err(MerkleTreeError::InvalidProof);
+                    }
+                    proof_pointers[i] += 1;
+                    self.nodes[i][pointer]
+                };
+
+                // get the node from the map of hashed nodes
+                let node = match v.get(&node_index) {
+                    Some(node) => node,
+                    None => return Err(MerkleTreeError::InvalidProof),
+                };
+
+                // compute parent node from node and sibling
+                partial_tree_map.insert(node_index ^ 1, sibling);
+                let parent = if node_index & 1 != 0 {
+                    H::merge(&[sibling, *node])
+                } else {
+                    H::merge(&[*node, sibling])
+                };
+
+                // add the parent node to the next set of nodes and partial_tree
+                let parent_index = node_index >> 1;
+                v.insert(parent_index, parent);
+                next_indexes.push(parent_index);
+                partial_tree_map.insert(parent_index, parent);
+
+                i += 1;
+            }
+        }
+
+        let result: Result<Vec<Vec<H::Digest>>, MerkleTreeError> = original_indexes
+            .iter()
+            .map(|&i| get_path::<H>(i, &partial_tree_map, self.depth as usize))
+            .collect();
+
+        result
```

Comment:
nit: since we return the result immediately after computing it, we could simplify this as:
```Rust
original_indexes
    .iter()
    .map(|&i| get_path::<H>(i, &partial_tree_map, self.depth as usize))
    .collect()
```